lib/galaxy/webapps/galaxy/api/authnz.py

```diff
@@ -0,0 +1,140 @@
+        `state` and `code` arguments, among others.
+
+        :rtype:  string
+        :return: a success/error message as a result of handling the callback.
+        """
+        user = trans.user.username if trans.user is not None else 'anonymous'
+        if not bool(kwargs):
+            raise MessageException(
+                err_msg="Did not receive any information from the `{}` identity provider to complete user `{}` "
+                        "authentication flow. Please try again, and if the problem persists, contact the Galaxy "
+                        "instance admin. Also note that this endpoint is to receive authentication callbacks only, "
+                        "and should not be called/reached by a user.".format(provider, user))
+        if 'error' in kwargs:
+            raise MessageException(
+                err_msg="Failed to handle authentication callback from {}. Please try again, and if the problem "
+                        "persists, contact the Galaxy instance admin".format(provider),
+                provider_returned_error=kwargs.get('error', 'None'))
+
+        success, message, (redirect_url, user) = trans.app.authnz_manager.callback(provider,
+                                                                                   kwargs['state'],
+                                                                                   kwargs['code'],
+                                                                                   trans,
+                                                                                   login_redirect_url=url_for('/'))
+        if success is False:
+            raise MessageException(err_msg=message)
+        user = user if user is not None else trans.user
+        if user is None:
+            raise MessageException(
+                err_msg="An unknown error occurred when handling the callback from `{}` identity provider. "
+                        "Please try again, and if the problem persists, contact the admin of this Galaxy "
+                        "instance.".format(provider))
+        trans.handle_user_login(user)
+        return {"message": message}
+
+    @web.expose_api
+    @web.require_login("authenticate against the selected identity provider")
+    def disconnect(self, trans, provider, **kwargs):
+        """
+        Get /api/authnz/{provider}/disconnect
+            Disconnects the logged-in user from the given identity provider (`provider`).
+            For details, see the `disconnect` function in the following path `lib/authnz/psa_authnz.py`.
+
+        :type  trans: galaxy.web.framework.webapp.GalaxyWebTransaction
+        :param trans: Galaxy web transaction.
+
+        :type  provider: string
+        :param provider: The name of the provider which should be used to authenticate a user.
+
+        :param kwargs:
+
+        :rtype:
+        :return:
+        """
+        if trans.user is None:
```

Comment:
If you want to check this, I'd just do ``assert trans.user``. This has a ``require_login`` already, so this isn't going to be reached right? Thus an assertion as a sanity check should be used if anything at all.